annotation_pipeline/image.py

```diff
@@ -7,35 +7,96 @@
+        self._partition = 0
+
+    def __call__(self, f_i, f_metrics, f_images):
+        self.add_f_metrics(f_i, f_metrics)
+        self.add_f_images(f_i, f_images)
+
+    @staticmethod
+    def images_size(f_images):
+        return sum(img.data.nbytes + img.row.nbytes + img.col.nbytes for img in f_images if img is not None)
+
+    def add_f_images(self, f_i, f_images):
+        self.formula_images[f_i] = f_images
+        self._formula_images_size += ImagesManager.images_size(f_images)
+        if self._formula_images_size > self.__class__.max_formula_images_size:
+            self.save_images()
+            self.formula_images.clear()
+            self._formula_images_size = 0
+
+    def add_f_metrics(self, f_i, f_metrics):
+        self.formula_metrics[f_i] = f_metrics
+
+    def save_images(self):
+        if self.formula_images:
+            print(f'Saving {len(self.formula_images)} images')
+            self._ibm_cos.put_object(Bucket=self._bucket,
+                                     Key=f'{self._prefix}/{self._partition}.pickle',
+                                     Body=pickle.dumps(self.formula_images))
+            self._partition += 1
+        else:
+            print(f'No images to save')
 
+    def finish(self):
+        self.save_images()
+        self.formula_images.clear()
+        self._formula_images_size = 0
 
-def gen_iso_images(sp_inds, sp_mzs, sp_ints, centr_df, nrows, ncols, ppm=3, min_px=1):
-    if len(sp_inds) > 0:
-        by_sp_mz = np.argsort(sp_mzs)  # sort order by mz ascending
-        sp_mzs = sp_mzs[by_sp_mz]
-        sp_inds = sp_inds[by_sp_mz]
-        sp_ints = sp_ints[by_sp_mz]
 
-        by_centr_mz = np.argsort(centr_df.mz.values)  # sort order by mz ascending
-        centr_mzs = centr_df.mz.values[by_centr_mz]
-        centr_f_inds = centr_df.formula_i.values[by_centr_mz]
-        centr_p_inds = centr_df.peak_i.values[by_centr_mz]
-        centr_ints = centr_df.int.values[by_centr_mz]
+def gen_iso_images(sp_inds, sp_mzs, sp_ints, centr_f_inds, centr_p_inds, centr_mzs, centr_ints, nrows, ncols, ppm=3, min_px=1):
+    # assume sp data is sorted by mz order ascending
+    # assume centr data is sorted by mz order ascending
 
+    def yield_buffer(buffer):
+        while len(buffer) < ISOTOPIC_PEAK_N:
+            buffer.append((buffer[0][0], len(buffer) - 1, 0, None))
+        buffer = np.array(buffer)
+        buffer = buffer[buffer[:, 1].argsort()]  # sort order by peak ascending
+        f_i = buffer[0][0]
+        f_ints = buffer[:, 2]
+        f_images = buffer[:, 3]
+        return f_i, f_ints, f_images
+
+    if len(sp_inds) > 0:
         lower = centr_mzs - centr_mzs * ppm * 1e-6
         upper = centr_mzs + centr_mzs * ppm * 1e-6
         lower_idx = np.searchsorted(sp_mzs, lower, 'l')
         upper_idx = np.searchsorted(sp_mzs, upper, 'r')
+        ranges_df = pd.DataFrame({'formula_i': centr_f_inds, 'range': zip(lower_idx, upper_idx)}).sort_values('formula_i')
```

Comment:
It would be better if `lower_idx` and `upper_idx` were separate columns. `zip`ping them allocates a bunch of extra memory for tuples and Python-ints. If you pass them separately, Pandas can keep them in numpy arrays.

Reply:
changed